pkg/toolset/snap_gen.go

```diff
@@ -15,12 +18,78 @@ import (
 	iotago "github.com/iotaledger/iota.go/v3"
 )
 
+type AddressWithBalance struct {
+	Address iotago.Address
+	Balance uint64
+}
+
+type jsonGenesisAddresses struct {
+	Balances map[string]uint64 `json:"balances"`
+	Rewards  map[string]uint64 `json:"rewards"`
+}
+
+type GenesisAddresses struct {
+	Balances []*AddressWithBalance
+}
+
+func (g *GenesisAddresses) MarshalJSON() ([]byte, error) {
+	return nil, nil
+}
+
+func (g *GenesisAddresses) UnmarshalJSON(bytes []byte) error {
+	j := &jsonGenesisAddresses{}
+	if err := json.Unmarshal(bytes, j); err != nil {
+		return err
+	}
+
+	if len(j.Balances) != 0 && len(j.Rewards) != 0 {
+		return errors.New("cannot specify both balances and rewards")
+	}
+
+	parseBalance := func(bech32Address string, balance uint64) error {
+		_, address, err := iotago.ParseBech32(bech32Address)
+		if err != nil {
+			if len(bech32Address) != 64 {
+				return err
+			}
+
+			// try parsing as hex
+			address, err = iotago.ParseEd25519AddressFromHexString("0x" + bech32Address)
+			if err != nil {
+				return err
+			}
+		}
+
+		g.Balances = append(g.Balances, &AddressWithBalance{
+			Address: address,
+			Balance: balance,
+		})
+
+		return nil
+	}
+
+	for bech32Address, balance := range j.Balances {
+		if err := parseBalance(bech32Address, balance); err != nil {
+			return err
+		}
+	}
+
+	for bech32Address, balance := range j.Rewards {
+		if err := parseBalance(bech32Address, balance); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 func snapshotGen(args []string) error {
 
 	fs := configuration.NewUnsortedFlagSet("", flag.ContinueOnError)
 	protocolParametersPathFlag := fs.String(FlagToolProtocolParametersPath, "", "the path to the initial protocol parameters file")
 	mintAddressFlag := fs.String(FlagToolSnapGenMintAddress, "", "the initial ed25519 address all the tokens will be minted to")
 	treasuryAllocationFlag := fs.Uint64(FlagToolSnapGenTreasuryAllocation, 0, "the amount of tokens to reside within the treasury, the delta from the supply will be allocated to 'mintAddress'")
+	genesisAddressesPathFlag := fs.String(FlagToolGenesisAddressesPath, "", "the file path to the genesis addresses file (optional)")
```

Comment:
maybe it would be good to be able to pass 2 files. Just thinking e.g. for Shimmer where we want to pass the unmodified staking rewards file and an additional file for the supply increase